rust/sbd-client/src/raw_client.rs

```diff
@@ -0,0 +1,303 @@
+        self.send.flush().await.map_err(Error::other)?;
+        Ok(())
+    }
+
+    /// Close the connection.
+    pub async fn close(&mut self) {
+        use futures::sink::SinkExt;
+        let _ = self.send.close().await;
+    }
+}
+
+/// The receive half of the websocket connection.
+pub struct WsRawRecv {
+    recv: WsRecv,
+}
+
+impl WsRawRecv {
+    /// Receive from the websocket.
+    pub async fn recv(&mut self) -> Result<Vec<u8>> {
+        use futures::stream::StreamExt;
+        use tokio_tungstenite::tungstenite::protocol::Message::*;
+        loop {
+            match self.recv.next().await {
+                None => return Err(Error::other("closed")),
+                Some(r) => {
+                    let msg = r.map_err(Error::other)?;
+                    match msg {
+                        Text(s) => return Ok(s.into_bytes()),
+                        Binary(v) => return Ok(v),
+                        Ping(_) | Pong(_) => (),
+                        Close(_) => return Err(Error::other("closed")),
+                        Frame(_) => unreachable!(),
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Process the standard sbd handshake from the client side.
+pub struct Handshake {
+    /// limit_byte_nanos.
+    pub limit_byte_nanos: i32,
+
+    /// limit_idle_millis.
+    pub limit_idle_millis: i32,
+
+    /// bytes sent.
+    pub bytes_sent: usize,
+}
+
+impl Handshake {
+    /// Process the standard sbd handshake from the client side.
+    pub async fn handshake<C: Crypto>(
+        send: &mut WsRawSend,
+        recv: &mut WsRawRecv,
+        crypto: &C,
+    ) -> Result<Self> {
+        let mut limit_byte_nanos = 8000;
+        let mut limit_idle_millis = 10_000;
+        let mut bytes_sent = 0;
+
+        loop {
+            match Msg(recv.recv().await?).parse()? {
+                MsgType::Msg { .. } => {
+                    return Err(Error::other("invalid handshake"))
+                }
+                MsgType::LimitByteNanos(l) => limit_byte_nanos = l,
+                MsgType::LimitIdleMillis(l) => limit_idle_millis = l,
+                MsgType::AuthReq(nonce) => {
+                    let sig = crypto.sign(nonce);
+                    let mut auth_res = Vec::with_capacity(32 + 64);
+                    auth_res.extend_from_slice(CMD_FLAG);
+                    auth_res.extend_from_slice(b"ares");
+                    auth_res.extend_from_slice(&sig);
+                    send.send(auth_res).await?;
+                    bytes_sent += 32 + 64;
+                }
+                MsgType::Ready => break,
+                MsgType::Unknown => (),
+            }
+        }
+
+        Ok(Self {
+            limit_byte_nanos,
+            limit_idle_millis,
+            bytes_sent,
+        })
+    }
+}
+
+fn priv_system_tls(
+    danger_disable_certificate_check: bool,
+) -> Arc<rustls::ClientConfig> {
+    let mut roots = rustls::RootCertStore::empty();
+
+    #[cfg(not(any(
```

Comment:
This is anticipating WASM where there won't be a system dir with root certs?

Reply:
Actually for mobile. The native-certs library doesn't support mobile yet, so for android/ios we build in the root certs.